### ISSUE TITLE
Support for writing preloaded data

### ISSUE THREAD

**Issue by hoechenberger** [mne-tools/mne-bids#714]
I advised @MaelysSolal to store some preprocessed data in the BIDS format as a derivative dataset. However, this is super awkward using `write_raw_bids`, as it won't allow us to store loaded data. So we now have to save everything to a temporary file, which we then have to read again (but without preloading), so we can write it using MNE-BIDS.

Two propositions:

- either add a kwarg to `write_raw_bids()` which allow us to store preloaded data
- or we could add `write_derivative()` that doesn't insist on data not being preloaded

WDYT?

cc @jasmainak @sappelhoff @agramfort

**Comment by agramfort** [comment]
would it be black magic to detect that the word derivatives is in the bids_root to allow preloaded data?
yet for @MaelysSolal you should write epochs for the "windows" of EEG data.

**Comment by hoechenberger** [comment]
> would it be black magic to detect that the word derivatives is in the bids_root to allow preloaded data?

That would be one possibility. We could also have a kwarg, `is_derivative=True` maybe?

> yet for @MaelysSolal you should write epochs for the "windows" of EEG data.

We were just looking at the resampled & filtered continuous data here, but yes, Epochs would potentially make sense depending on how much time the windowing takes

**Comment by adam2392** [comment]
Should we just add all the entities on this list? https://bids-specification.readthedocs.io/en/stable/99-appendices/09-entities.html We wouldn't support IO for each type of file (e.g. PET scans), but BIDSPath can easily be used as an object to traverse any type of BIDS dataset that way?

### Some notes from my understanding
To define a derivative dataset, we would need:

- the BIDS root: e.g. `bids_root/`
- "derivative chain": e.g. `resampled/preproc1/preproc2/`

and then any dataset could theoretically be written in the following manner:

`<root>/derivatives/<deriv_chain>/<subject/session/datatype directory structure>/`

explicitly:

`bids_root/derivatives/resampled/preproc1/preproc2/sub-x/ses-y/ieeg/*`

Another set of questions based on the [spec](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets). 

1. There are "derivative" chains, is there anyway to easily abstract this? If so, most of the write framework can be easily done in the same manner?
2. When writing derivatives though we don't necessarily need all the "sidecar" metadata. Simply a trace back to the source raw dataset, which can then be used to populate metadata for the derivative?
3. Safe to say must use `fif`, or `vhdr` for mne-python related derivatives at least?

**Comment by hoechenberger** [comment]
> Should we just add all the entities on this list?

I don't see how this is specifically related to derivatives..?



> and then any dataset could theoretically be written in the following manner:
> 
> `<root>/derivatives/<deriv_chain>/<subject/session/datatype directory structure>/`
> 
> explicitly:
> 
> `bids_root/derivatives/resampled/preproc1/preproc2/sub-x/ses-y/ieeg/*`

I'm not sure what BIDS says on this, but I already anticipate that @agramfort and I will need to be able to store derivatives someplace outside the original BIDS root when dealing with certain large datasets. We've managed to fill up hundreds of GBs of storage in no time in the past.






> Another set of questions based on the [spec](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets).
> 
> 1. There are "derivative" chains, is there anyway to easily abstract this? If so, most of the write framework can be easily done in the same manner?
> 2. When writing derivatives though we don't necessarily need all the "sidecar" metadata. Simply a trace back to the source raw dataset, which can then be used to populate metadata for the derivative?
> 3. Safe to say must use `fif`, or `vhdr` for mne-python related derivatives at least?

A bunch of BIDS-heads including recently had a video conference to start discussing electrophysiological derivatives. This is still in a super early stage and I expect the specs to change. Therefore, I'd keep things as simple as possible for now. 

I'd suggest for starters we only support writing continuous data, which would allow us to reuse the existing machinery from `write_raw_bids()`, including the sidecars. Data could be filtered, decimated, annotated etc – first steps of preprocessing. Implementing anything beyond that is not going to bring us much joy at this stage.

**Comment by agramfort** [comment]
in the team we don't always store derivatives in the bids_root/derivatives
folder for 2 reasons:
- bids_root is read only for most users
- derivatives can be different for each user although the name may conflict

any other suggestion besides the is_derivative param?

thoughts everyone?

>

**Comment by hoechenberger** [comment]
> any other suggestion besides the is_derivative param?

I was thinking about this and I'm now leaning towards introducing a new function, `write_derivative` or `write_derivative_bids`, or `write_bids_derivative`, that takes an instance of data. For now it would only work with Raw objects and internally call `write_raw_bids`

The reasons why I'm thinking a new function would be better than simply adding a kwarg to `write_raw_bids`:
- `write_derivative` would be more generic and could be extended, one by one, to support more data types than just Raw
- semantics would be cleaner: `write_raw_bids` is supposed to write **raw** data, while derivative continuous data would be **continuous** data of the `mne.io.Raw` type, but not actual "raw" data

**Comment by agramfort** [comment]
not sure. raw in mne means continuous data. We use everywhere the word raw
for derivatives.

with only one function you would just call inst.save(bids_path.fpath) + the
sidecar files?

>

**Comment by hoechenberger** [comment]
> not sure. raw in mne means continuous data. We use everywhere the word raw for derivatives. 

Yes, that's what we're doing now in the Study Template (which I think is what you're referring to here?). Just as a recap for others following this discussion, in the Study Template file names might evolve like this, goingraw data over Maxwell filtering to frequency filtering:

- raw: `sub-01_task-foo_meg.fif`
- Maxwell-filtered: `sub-01_task-foo_proc-sss_raw.fif`
- frequency-filtered: `sub-01_task-foo_proc-filt_raw.fif`

Epochs would get the suffix `_epo` etc. to be compatible with MNE-Python naming.

This is a naming scheme **we** cooked up, as BIDS doesn't specify this yet.

In BIDS, "raw" has a different meaning: it [refers](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#source-vs-raw-vs-derived-data) to unprocessed or only minimally processed data; the continuous data we produce e.g. by filtering the raw data would be called a **derivative of the raw data** in BIDS, even though it would be represented by the exact same data type in MNE (`Raw`). I think we need to be very careful not to get mixed up here.

Another question we should discuss with other BIDS folks is how to generate some sort of "provenance" of derivative generation. For example, I could imagine having a JSON file that keeps track of the input data used to generate a certain derivative:

```json
{
    "sub-01_task-foo_proc-sss_raw.fif":{
        "sourceFileName": "sub-01_task-foo_meg.fif",
        "sourceFileHash": "12345",
        "hashAlgorithm": "MD5"
    }
}
```

> with only one function you would just call inst.save(bids_path.fpath) + the sidecar files?

Basically for now my idea would be to move the implementation of `write_raw_bids()` to a private function `_write_raw_bids()` that will be called by `write_raw_bids()`, and can also be used by a future `write_derivative()` if a `Raw` object needs to be written. Necessary adjustments to the sidecars could then be done inside this function too.

**Comment by agramfort** [comment]
and the write_derivative would do the branching between type (BaseRaw,
BaseEpochs etc.)
to know what and how to save?

>

**Comment by hoechenberger** [comment]
> and the write_derivative would do the branching between type (BaseRaw, BaseEpochs etc.) to know what and how to save?

Yes, exactly!

**Comment by jasmainak** [comment]
read quickly but I'd suggest a different `write_*` function for each derivative type. To not creating a mess by overloading one function. It's a pet peeve of mine how python folks just like to overload one function with 20 parameters ;-)

**Comment by hoechenberger** [comment]
This has been resolved via #819
